### PR TITLE
Templatecpp

### PR DESCRIPTION
**Pull request by kmelmon**
This change adds support to the CLI for creating a C++ based template app.  The CLI will continue to support C# through a --language parameter, but C++ will be the default.

The C++ template is basically a port of the C# template to C++.

Added verification of both types of template apps to the CI loop.

Issue tracked by #2337.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3262)

### CODE REVIEW

**Comment by kmelmon** [comment]
My CI loop hit the build issue in Folly that we're currently discussing on a thread.  Other than that the CI loop looks clean.

File: .ado/templates/react-native-init.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  - task: CmdLine@2
+    displayName: Install react-native cli
+    inputs:
+      script: npm install -g react-native-cli
+
+  - task: CmdLine@2
+    displayName: Init new project
+    inputs:
+      script: react-native init testcli --version 0.60.6
```
Comment (JunielKatarn) [inline .ado/templates/react-native-init.yml]: Parameterize version.

File: .ado/templates/react-native-init.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    inputs:
+      command: restore
+      restoreSolution: $(Agent.BuildDirectory)\testcli\windows\testcli.sln
+
+  - task: MSBuild@1
+    displayName: MSBuild - testcli
+    inputs:
+      solution: $(Agent.BuildDirectory)\testcli\windows\testcli.sln
+      msbuildVersion: $(MSBuildVersion) # Optional. Options: latest, 16.0, 15.0, 14.0, 12.0, 4.0
+      msbuildArchitecture: x86 # Optional. Options: x86, x64
```
Comment (JunielKatarn) [inline .ado/templates/react-native-init.yml]: Why not x64?

File: .ado/templates/react-native-init.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      platform: x64 # Optional
```
Comment (JunielKatarn) [inline .ado/templates/react-native-init.yml]: Parameterize platform and configuration.

Is this meant to only build for `x64|Debug`?

File: .ado/templates/react-native-init.yml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    inputs:
+      script: yarn build
+      workingDirectory: vnext
+
+  # yarn ends up copying the whole node_modules folder when doing an install of a file package
+  # Delete node_modules, so that resolution is more like when installing from a published npm package
+  - task: CmdLine@2
+    displayName: Remove node_modules
+    inputs:
+      script: rd /S /Q node_modules
```
Comment (JunielKatarn) [inline .ado/templates/react-native-init.yml]: Why is this needed?

File: .ado/templates/react-native-init.yml [hunk 5]
```diff
@@ -0,0 +1,80 @@
+#
```
Comment (JunielKatarn) [inline .ado/templates/react-native-init.yml]: Steps from `yarn install` to `MSBuild` seem repetitive.
Why not use `templates/build-rnw` within this template instead?

File: .ado/templates/react-native-init.yml (hunk 2 above)
Comment (kmelmon) [reply .ado/templates/react-native-init.yml]: I don't know - it was this way before my change.  I'm trying an experiment now with x64.

---
In reply to: [328923375](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r328923375) [](ancestors = 328923375)

**Comment by kmelmon** [comment]
re-running passed.  weird.

---
In reply to: [535778183](https://github.com/microsoft/react-native-windows/pull/3262#issuecomment-535778183) [](ancestors = 535778183)

File: .ado/templates/react-native-init.yml (hunk 1 above)
Comment (kmelmon) [reply .ado/templates/react-native-init.yml]: done

---
In reply to: [328923250](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r328923250) [](ancestors = 328923250)

File: .ado/templates/react-native-init.yml (hunk 3 above)
Comment (kmelmon) [reply .ado/templates/react-native-init.yml]: parameters added.  Yes this is just building x64 debug, to save time.  We're already the framework in multiple flavors.

---
In reply to: [328923495](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r328923495) [](ancestors = 328923495)

File: .ado/templates/react-native-init.yml (hunk 4 above)
Comment (kmelmon) [reply .ado/templates/react-native-init.yml]: This code was already here, see comments above.  I'm opting to keep it this way.

---
In reply to: [328923793](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r328923793) [](ancestors = 328923793)

File: .ado/templates/react-native-init.yml (hunk 5 above)
Comment (kmelmon) [reply .ado/templates/react-native-init.yml]: The steps are somewhat unique - it needs to install the CLI and RNW from a file path, then build the CLI app.

---
In reply to: [328924431](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r328924431) [](ancestors = 328924431)

File: .ado/templates/react-native-init.yml (hunk 2 above)
Comment (kmelmon) [reply .ado/templates/react-native-init.yml]: the experiment failed.  the C# template app won't launch when built when x64 tools - some assembly is missing from the build machine.  Really we shouldn't be launching test apps on the build machines, and we will be addressing this for our e2e tests.  For now I've gone back to using x86 build tools.

---
In reply to: [329257330](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r329257330) [](ancestors = 329257330,328923375)

File: current/local-cli/rnpm/windows/index.js [hunk 6]
```diff
@@ -15,6 +15,10 @@ module.exports = [{
   }, {
     command: '--template [template]',
     description: 'Template to install. E.g., `vnext`.',
+  }, {
+    command: '--language [language]',
+    description: 'Which language the template app is written in.  Possible values are cs and cpp',
```
Comment (licanhua) [inline current/local-cli/rnpm/windows/index.js]: > Possible [](start = 65, length = 9)

Remove possible

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.xaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:<%=ns%>"
+    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+      xmlns:react="using:react.uwp"
+    mc:Ignorable="d"
+    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
+
+    <Grid>
```
Comment (licanhua) [inline vnext/local-cli/generator-windows/templates/cpp/src/MainPage.xaml]: >   [](start = 2, length = 2)

nit: remove two spaces

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
+﻿#include "pch.h"
+#include "MainPage.h"
+#include "MainPage.g.cpp"
+
+using namespace winrt;
+using namespace Windows::UI::Xaml;
+using namespace react::uwp;
+
+// clang-format off
```
Comment (licanhua) [inline vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp]: customer doesn't need this.

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace winrt::<%=ns%>::implementation
+{
+	const wchar_t* JSFILENAME = L"index";
+	const wchar_t* JSCOMPONENTNAME = L"<%=name%>";
+
+    MainPage::MainPage()
+    {
+        InitializeComponent();
+		LoadReact();
```
Comment (licanhua) [inline vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp]: >		 [](start = 0, length = 2)

nit: please use space. it applies to other .h and .cpp files too

File: current/local-cli/rnpm/windows/index.js [hunk 10]
```diff
@@ -15,6 +15,10 @@ module.exports = [{
   }, {
     command: '--template [template]',
     description: 'Template to install. E.g., `vnext`.',
+  }, {
+    command: '--language [language]',
```
Comment (licanhua) [inline current/local-cli/rnpm/windows/index.js]: You are make change to current. do we need to copy rnpm-plugin-windows to vnext and publish as next package too?

Comment (kmelmon) [reply current/local-cli/rnpm/windows/index.js]: This code is shared between current and vnext.  What is needed is to simply publish a new rnpm-plugin-windows.  @marlenecota is this done manually or automatically?

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.xaml (hunk 7 above)
Comment (kmelmon) [reply vnext/local-cli/generator-windows/templates/cpp/src/MainPage.xaml]: fixed

---
In reply to: [329279850](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r329279850) [](ancestors = 329279850)

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp (hunk 8 above)
Comment (kmelmon) [reply vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp]: This is used to get past an issue with yarn format mucking with the template variables.  I added logic to strip these out as the template is being copied (see vnext\generator-windows\index.js) so the developer won't see this when they run the CLI.

---
In reply to: [329280236](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r329280236) [](ancestors = 329280236)

File: vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp (hunk 9 above)
Comment (kmelmon) [reply vnext/local-cli/generator-windows/templates/cpp/src/MainPage.cpp]: good catch, fixed.

---
In reply to: [329280361](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r329280361) [](ancestors = 329280361)

File: current/local-cli/rnpm/windows/index.js (hunk 6 above)
Comment (kmelmon) [reply current/local-cli/rnpm/windows/index.js]: I prefer this language.

---
In reply to: [329279587](https://github.com/microsoft/react-native-windows/pull/3262#discussion_r329279587) [](ancestors = 329279587)

File: current/local-cli/rnpm/windows/index.js (hunk 10 above)
Comment (marlenecota) [reply current/local-cli/rnpm/windows/index.js]: This is published automatically.
